### DISCUSSION TITLE
Environment variables can't be sourced indirectly

### DISCUSSION THREAD

**Discussion by eakl** [vercel/next.js#36696] [Help] [answered]
I have an issue using environment variables with NextJS.

Basically I collect all ENV variable in a config file and use this config file in my project.
In this case I can't source the config file in my util function but I can source the env variable directly.

What's wrong?

**Project structure**
```
|__ config
|    |__ index.js
|
|__ pages
|    |__ api
|    |    |__ formHandler.js
|    |
|    |__ page.js
|
|__ utils
    |__ index.js
```

**config/index.js**
```javascript
export const getEnvVar = (envVar) => {
  const unvalidatedEnvVar = process.env[envVar]

  if (!unvalidatedEnvVar) {
    throw new Error(`Couldn't find environment variable: ${envVar}`)
  } else {
    return unvalidatedEnvVar
  }
}

export const config = {
  PREFIX: getEnvVar('MY_SECRET_PREFIX'),
  STRING: getEnvVar('MY_SECRET_STRING'),
}
```

**utils/index.js**
```javascript
import { config } from '@config'

export const getPrefixedString = (string) => {
  const prefix = config.PREFIX // THIS causes getEnvVar to throw
  // const prefix = process.env.MY_SECRET_PREFIX // THIS works
  return `${prefix}_${string}`
}
```

**page/api/formHandler.js**
```javascript
import { config } from '@config'
import { getPrefixedString } from '@utils'

const prefixedString = getPrefixedString(config.STRING)

export default async (req, res) => {
  return res.status(200).json({ string: prefixedString })
}

```

**page/page.js**
```javascript
// ... a form component which onSubmit calls the formHandler api throw fetch('/api/formHandler')
```

**Comment by icyJoseph** [comment, marked as answer]
~IIRC, you must statically define each key of `process.env` at build time, otherwise, it won't work.~

The [docs](https://nextjs.org/docs/basic-features/environment-variables) say:

> Note: In order to keep server-only secrets safe, Next.js replaces process.env.* with the correct values at build time. This means that process.env is not a standard JavaScript object, so you’re not able to use [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment). Environment variables must be referenced as e.g. process.env.PUBLISHABLE_KEY, not const { PUBLISHABLE_KEY } = process.env.

~I think the hint at not being a standard JavaScript object, is perhaps not strong enough, but I'd say that in your case, you could try to statically define the variables you need and give it a go again.~

In a project I maintain we have a `getServerConfig` file, (which apparently I implemented) and there I have statically defined what I need from `process.env`, but I seem to remember that dynamic access gave me issues.

---

### Edit

I've done an experiment:

```ts
// config/getConfig.ts
export const getConfig = (key: string) => {
  const maybeKey = process.env[key];

  if (!maybeKey) {
    throw new Error(`Key: ${key} is not defined in process.env`);
  }

  return maybeKey;
};
```

And then:

```ts
// api/hello.ts
import type { NextApiRequest, NextApiResponse } from "next";

import { getConfig } from "../../config/getConfig";

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const { query } = req;

  const { key } = query;

  if (key && !Array.isArray(key)) {
    const value = getConfig(key);
    console.log({ value }); // works! 
  }

  res.status(200).json({ name: "John Doe" });
}
```

And these env's:

```
X=1
Y=2
Z=3
W=4
```

And then sending these two requests:

```shell
###
GET http://localhost:3000/api/hello?key=X

###
GET http://localhost:3000/api/hello?key=P
```

Then the first one works, and the second throws. So it should work to access them dynamically. 

It must be something in the way your project is organised.

**Perhaps a client side module, imports a type or a constant from a file that's has dependency on `getConfig`?**

**Reply by eakl** [reply]
The thing is I can access my config file in many places in my project. For example in my `lib` folder which contain CMS, Database, Email provider clients that are used in server-side functions.

there, `const API_KEY = config.EMAIL_PROVIDER.API_KEY` works normally.

So I guess it comes from the way and the place I call my config file. More than the NextJS fashion of handling process.env object

**Comment by GabenGar** [comment]
If you want store the env values in variables, you can create a separate module where you do calls to all needed `process.env` keys and export the result of each call. This way consumer modules can get the values by importing the variables.

**Reply by eakl** [reply]
One function that returns all `process.env.XXX` at each call?

**Reply by icyJoseph** [reply]
He means something like:

```js
export const dbURL = process.env.DATABASE_URL;
// and a bunch more
// or

export const getDBUrl = () => process.env.DATABASE_URL;
```

And then in anywhere in your app you can:

```js
import {getDBUrl, dbURL} from 'lib/config';
```

At least, that's what I got. You could even group the exports by environment and such.

**Reply by GabenGar** [reply]
@eakl 
No, one module which calls `process.env.XXX` for each key and exports the values as named exports. The bonus of this approach is NextJS inlines `process.env` values at build time and then webpack code splits them apart, so every lambda will only get the env vars values they use in their code instead of dragging the entire config around.

**Comment by eakl** [comment]
For example

```javascript
export const getPrefixedString = async (string) => {
  const { config } = await import('@config') // THIS works
  const prefix = config.PREFIX // and THIS works
  return `${prefix}_${string}`
}
```

but I don't understand why

**Reply by icyJoseph** [reply]
`await` without `async`? Perhaps you have a repo to look at? Or perhaps you could prepare a small reproduction repo?

**Reply by eakl** [reply]
My bad, the function is `async`. I mean, calling my config file with dynamic import seems to work but not importing the config file statically at the top of the `getPrefixedString` module.

**Reply by icyJoseph** [reply]
Aha, please see my first comment, with updates.

**Comment by eakl** [comment]
My `@utils` module is organized as follows
```
|__ utils
    |__ index.js
    |__ utilFunction1.js
    |__ utilFunction2.js
    |__ getPrefixedString.js
    |__ ...
```

And indeed I import `@utils` in client side modules to get access to `utilFunctionX.ts`

I tried to access `@config` in other util functions, it throw as well. I guess NextJS prevents the entire `@utils` module to access the config file as this module is called client-side.
